File: csharp/Prescribe.Core/Runtime/StdLib.cs
using Prescribe.Core.Diagnostics;
using Prescribe.Core.Semantics;
using Prescribe.Core.Util;

namespace Prescribe.Core.Runtime;

public sealed class StdLib
{
    private long _seed = 1;

    public RuntimeValue Rand()
    {
        _seed = (1103515245 * _seed + 12345) % 2147483648;
        return RuntimeValues.Make(Types.Real, _seed / 2147483648.0);
    }

    public RuntimeValue Length(RuntimeValue arg, int line)
    {
        if (arg.Type.Kind != TypeKind.String) throw Errors.At(ErrorType.TypeError, line, "LENGTH requires STRING.");
        return RuntimeValues.Make(Types.Integer, ((string)arg.Data!).Length);
    }

    public RuntimeValue Right(RuntimeValue s, RuntimeValue n, int line)
    {
        if (s.Type.Kind != TypeKind.String || n.Type.Kind != TypeKind.Integer) throw Errors.At(ErrorType.TypeError, line, "RIGHT requires STRING, INTEGER.");
        var str = (string)s.Data!;
        var count = (int)n.Data!;
        if (count < 0 || count > str.Length) throw Errors.At(ErrorType.RangeError, line, "RIGHT out of range.");
        return RuntimeValues.Make(Types.String, str.Substring(str.Length - count));
    }

    public RuntimeValue Mid(RuntimeValue s, RuntimeValue start, RuntimeValue n, int line)
    {
        if (s.Type.Kind != TypeKind.String || start.Type.Kind != TypeKind.Integer || n.Type.Kind != TypeKind.Integer)
        {
            throw Errors.At(ErrorType.TypeError, line, "MID requires STRING, INTEGER, INTEGER.");
        }
        var str = (string)s.Data!;
        var pos = (int)start.Data!;
        var count = (int)n.Data!;
        if (count < 0 || pos < 1 || pos > str.Length + 1 || pos + count - 1 > str.Length)
        {
            throw Errors.At(ErrorType.RangeError, line, "MID out of range.");
        }
        return RuntimeValues.Make(Types.String, str.Substring(pos - 1, count));
    }

    public RuntimeValue LCase(RuntimeValue s, int line)
    {
        if (s.Type.Kind != TypeKind.String) throw Errors.At(ErrorType.TypeError, line, "LCASE requires STRING.");
        return RuntimeValues.Make(Types.String, ToLowerAscii((string)s.Data!));
    }

    public RuntimeValue UCase(RuntimeValue s, int line)
    {
        if (s.Type.Kind != TypeKind.String) throw Errors.At(ErrorType.TypeError, line, "UCASE requires STRING.");
        return RuntimeValues.Make(Types.String, ToUpperAscii((string)s.Data!));
    }

    public RuntimeValue Int(RuntimeValue x, int line)
    {
        if (x.Type.Kind != TypeKind.Real) throw Errors.At(ErrorType.TypeError, line, "INT requires REAL.");
        return RuntimeValues.Make(Types.Integer, (int)Math.Truncate((double)x.Data!));
    }

    public RuntimeValue Real(RuntimeValue x, int line)
    {
        if (x.Type.Kind != TypeKind.Integer) throw Errors.At(ErrorType.TypeError, line, "REAL requires INTEGER.");
        return RuntimeValues.Make(Types.Real, Convert.ToDouble(x.Data));
    }

    public RuntimeValue String(RuntimeValue x, int line)
    {
        return RuntimeValues.Make(Types.String, ToOutputString(x, line));
    }

    public RuntimeValue Char(RuntimeValue x, int line)
    {
        if (x.Type.Kind != TypeKind.Integer) throw Errors.At(ErrorType.TypeError, line, "CHAR requires INTEGER.");
        var num = (int)x.Data!;
        if (num < 0 || num > 127) throw Errors.At(ErrorType.RangeError, line, "CHAR out of range.");
        return RuntimeValues.Make(Types.Char, ((char)num).ToString());
    }

    public RuntimeValue Boolean(RuntimeValue x, int line)
    {
        if (x.Type.Kind != TypeKind.String) throw Errors.At(ErrorType.TypeError, line, "BOOLEAN requires STRING.");
        var text = ((string)x.Data!).ToUpperInvariant();
        if (text == "TRUE") return RuntimeValues.Make(Types.Boolean, true);
        if (text == "FALSE") return RuntimeValues.Make(Types.Boolean, false);
        throw Errors.At(ErrorType.TypeError, line, "Invalid BOOLEAN token.");
    }

    public RuntimeValue Date(RuntimeValue x, int line)
    {
        if (x.Type.Kind != TypeKind.String) throw Errors.At(ErrorType.TypeError, line, "DATE requires STRING.");
        var value = DateUtil.ParseDateLiteral((string)x.Data!, line);
        return RuntimeValues.Make(Types.Date, value);
    }

    public RuntimeValue Ord(RuntimeValue x, int line)
    {
        if (x.Type.Kind != TypeKind.Enum) throw Errors.At(ErrorType.TypeError, line, "ORD requires ENUM.");
        return RuntimeValues.Make(Types.Integer, (int)x.Data!);
    }

    public RuntimeValue EnumValue(TypeSymbol enumType, RuntimeValue ordinal, int line)
    {
        if (ordinal.Type.Kind != TypeKind.Integer) throw Errors.At(ErrorType.TypeError, line, "ENUMVALUE ordinal must be INTEGER.");
        var value = (int)ordinal.Data!;
        if (enumType is not EnumType enumTypeSym) throw Errors.At(ErrorType.TypeError, line, "ENUMVALUE requires enum type.");
        if (value < 0 || value >= enumTypeSym.Members.Count) throw Errors.At(ErrorType.RangeError, line, "ENUMVALUE out of range.");
        return RuntimeValues.Make(enumType, value);
    }

    public RuntimeValue Size(RuntimeValue setValue, int line)
    {
        if (setValue.Type.Kind != TypeKind.Set) throw Errors.At(ErrorType.TypeError, line, "SIZE requires SET.");
        return RuntimeValues.Make(Types.Integer, ((HashSet<int>)setValue.Data!).Count);
    }

    private static string ToLowerAscii(string value)
    {
        var chars = value.ToCharArray();
        for (var i = 0; i < chars.Length; i += 1)
        {
            if (chars[i] is >= 'A' and <= 'Z')
            {
                chars[i] = (char)(chars[i] + 32);
            }
        }
        return new string(chars);
    }

    private static string ToUpperAscii(string value)
    {
        var chars = value.ToCharArray();
        for (var i = 0; i < chars.Length; i += 1)
        {
            if (chars[i] is >= 'a' and <= 'z')
            {
                chars[i] = (char)(chars[i] - 32);
            }
        }
        return new string(chars);
    }

    public static string ToOutputString(RuntimeValue value, int line)
    {
        return value.Type.Kind switch
        {
            TypeKind.Integer => Convert.ToString(value.Data, System.Globalization.CultureInfo.InvariantCulture) ?? "0",
            TypeKind.Real => StringUtil.FormatReal(Convert.ToDouble(value.Data), line),
            TypeKind.Boolean => ((bool)value.Data!) ? "TRUE" : "FALSE",
            TypeKind.Char => (string)value.Data!,
            TypeKind.String => (string)value.Data!,
            TypeKind.Date => DateUtil.DateToString((DateValue)value.Data!),
            _ => throw Errors.At(ErrorType.TypeError, line, "Value not outputtable.")
        };
    }
}

File: legacy-typescript/src/runtime/stdlib.ts
import { errorAt } from "../diagnostics/errors.js";
import { Type } from "../semantics/types.js";
import { RuntimeValue } from "./values.js";
import { formatReal } from "../util/strings.js";
import { parseDateLiteral, dateToString } from "../util/dates.js";
import { makeValue } from "./values.js";

export class StdLib {
  private seed = 1;

  rand(): RuntimeValue {
    this.seed = (1103515245 * this.seed + 12345) % 2147483648;
    return makeValue({ kind: "Real" }, this.seed / 2147483648);
  }

  length(arg: RuntimeValue, line: number): RuntimeValue {
    if (arg.type.kind !== "String") throw errorAt("TypeError", line, "LENGTH requires STRING.");
    return makeValue({ kind: "Integer" }, (arg.data as string).length);
  }

  right(s: RuntimeValue, n: RuntimeValue, line: number): RuntimeValue {
    if (s.type.kind !== "String" || n.type.kind !== "Integer") throw errorAt("TypeError", line, "RIGHT requires STRING, INTEGER.");
    const str = s.data as string;
    const count = n.data as number;
    if (count < 0 || count > str.length) throw errorAt("RangeError", line, "RIGHT out of range.");
    return makeValue({ kind: "String" }, str.slice(str.length - count));
  }

  mid(s: RuntimeValue, start: RuntimeValue, n: RuntimeValue, line: number): RuntimeValue {
    if (s.type.kind !== "String" || start.type.kind !== "Integer" || n.type.kind !== "Integer") {
      throw errorAt("TypeError", line, "MID requires STRING, INTEGER, INTEGER.");
    }
    const str = s.data as string;
    const pos = start.data as number;
    const count = n.data as number;
    if (count < 0 || pos < 1 || pos > str.length + 1 || pos + count - 1 > str.length) {
      throw errorAt("RangeError", line, "MID out of range.");
    }
    return makeValue({ kind: "String" }, str.substr(pos - 1, count));
  }

  lcase(s: RuntimeValue, line: number): RuntimeValue {
    if (s.type.kind !== "String") throw errorAt("TypeError", line, "LCASE requires STRING.");
    return makeValue({ kind: "String" }, (s.data as string).replace(/[A-Z]/g, (c) => c.toLowerCase()));
  }

  ucase(s: RuntimeValue, line: number): RuntimeValue {
    if (s.type.kind !== "String") throw errorAt("TypeError", line, "UCASE requires STRING.");
    return makeValue({ kind: "String" }, (s.data as string).replace(/[a-z]/g, (c) => c.toUpperCase()));
  }

  int(x: RuntimeValue, line: number): RuntimeValue {
    if (x.type.kind !== "Real") throw errorAt("TypeError", line, "INT requires REAL.");
    return makeValue({ kind: "Integer" }, Math.trunc(x.data as number));
  }

  real(x: RuntimeValue, line: number): RuntimeValue {
    if (x.type.kind !== "Integer") throw errorAt("TypeError", line, "REAL requires INTEGER.");
    return makeValue({ kind: "Real" }, Number(x.data));
  }

  string(x: RuntimeValue, line: number): RuntimeValue {
    return makeValue({ kind: "String" }, toOutputString(x, line));
  }

  char(x: RuntimeValue, line: number): RuntimeValue {
    if (x.type.kind !== "Integer") throw errorAt("TypeError", line, "CHAR requires INTEGER.");
    const num = x.data as number;
    if (num < 0 || num > 127) throw errorAt("RangeError", line, "CHAR out of range.");
    return makeValue({ kind: "Char" }, String.fromCharCode(num));
  }

  boolean(x: RuntimeValue, line: number): RuntimeValue {
    if (x.type.kind !== "String") throw errorAt("TypeError", line, "BOOLEAN requires STRING.");
    const text = (x.data as string).toUpperCase();
    if (text === "TRUE") return makeValue({ kind: "Boolean" }, true);
    if (text === "FALSE") return makeValue({ kind: "Boolean" }, false);
    throw errorAt("TypeError", line, "Invalid BOOLEAN token.");
  }

  date(x: RuntimeValue, line: number): RuntimeValue {
    if (x.type.kind !== "String") throw errorAt("TypeError", line, "DATE requires STRING.");
    const value = parseDateLiteral(x.data as string, line);
    return makeValue({ kind: "Date" }, value);
  }

  ord(x: RuntimeValue, line: number): RuntimeValue {
    if (x.type.kind !== "Enum") throw errorAt("TypeError", line, "ORD requires ENUM.");
    return makeValue({ kind: "Integer" }, x.data as number);
  }

  enumValue(enumType: Type, ordinal: RuntimeValue, line: number): RuntimeValue {
    if (ordinal.type.kind !== "Integer") throw errorAt("TypeError", line, "ENUMVALUE ordinal must be INTEGER.");
    const value = ordinal.data as number;
    if (enumType.kind !== "Enum") throw errorAt("TypeError", line, "ENUMVALUE requires enum type.");
    if (value < 0 || value >= enumType.members.length) throw errorAt("RangeError", line, "ENUMVALUE out of range.");
    return makeValue(enumType, value);
  }

  size(setValue: RuntimeValue, line: number): RuntimeValue {
    if (setValue.type.kind !== "Set") throw errorAt("TypeError", line, "SIZE requires SET.");
    return makeValue({ kind: "Integer" }, (setValue.data as Set<number>).size);
  }
}

export function toOutputString(value: RuntimeValue, line: number): string {
  switch (value.type.kind) {
    case "Integer":
      return String(value.data as number);
    case "Real":
      return formatReal(value.data as number, line);
    case "Boolean":
      return (value.data as boolean) ? "TRUE" : "FALSE";
    case "Char":
      return value.data as string;
    case "String":
      return value.data as string;
    case "Date":
      return dateToString(value.data as any);
    default:
      throw errorAt("TypeError", line, "Value not outputtable.");
  }
}
